### ISSUE TITLE
DELETE Request spec wants number as first argument instead of url?

### ISSUE THREAD

**Issue by hmistry** [rails/rails#25102]
Hi,

### Steps to reproduce
I'm performing the following API request spec test using Rspec:
`delete "/api/notifications/unsubscribe", params: { device_token: device.device_token }, headers: headers(token: token)`

### Expected behavior
I expect a delete request to be performed (like my other get or post request tests) and the unsubscribe action to be performed.

### Actual behavior
I get the following error:
`*** "Delete" argument ""/api/notifications/unsubscribe"," needs to be a number`

I've checked the params/headers are correct, route does exist, and verified the route is working using curl. Has the argument params changed for delete request spec testing?

I posted this on Rspec [#1625](https://github.com/rspec/rspec-rails/issues/1625), but they mentioned it was a Rails command. I looked at the Rails doc and I believe my passed params in the method are correct i.e. `delete(path, *args)`... not sure where the "needs to be a number" is coming from.

### System configuration
**Rails version**:
Rails 5.0.0.rc1

**Ruby version**:
Ruby 2.3.0

**Comment by vipulnsward** [comment]
What's the trace for this error?

**Comment by hmistry** [comment]
So the above message is what I get debugging using `byebug`, no trace.

This is the trace what the test expectation fails with:
```
     # ./spec/requests/api/v0/notifications_spec.rb:85:in `block (3 levels) in <top (required)>'
     # /Users/hm/.rvm/gems/ruby-2.3.0@loop_server/gems/activesupport-5.0.0.rc1/lib/active_support/dependencies.rb:287:in `load'
     # /Users/hm/.rvm/gems/ruby-2.3.0@loop_server/gems/activesupport-5.0.0.rc1/lib/active_support/dependencies.rb:287:in `block in load'
     # /Users/hm/.rvm/gems/ruby-2.3.0@loop_server/gems/activesupport-5.0.0.rc1/lib/active_support/dependencies.rb:259:in `load_dependency'
     # /Users/hm/.rvm/gems/ruby-2.3.0@loop_server/gems/activesupport-5.0.0.rc1/lib/active_support/dependencies.rb:287:in `load'
     # /Users/hm/.rvm/gems/ruby-2.3.0@global/gems/spring-commands-rspec-1.0.4/lib/spring/commands/rspec.rb:18:in `call'

```

Let me know if you need anything else.

**Comment by sgrif** [comment]
The steps to reproduce that you've provided are not enough for us to reproduce this issue. Please provide a reproduction script using https://github.com/rails/rails/blob/master/guides/bug_report_templates/action_controller_master.rb as a template

**Comment by hmistry** [comment]
I'm creating the reproduction script but I'm getting a `ParamsParser` error when using Rspec, see below. This test worked with Minitest so I want to try with Rspec to see if it is the root cause of my issue.

Could you help me to get Rspec configured to work with the above template? I've replaced `minitest` section with the following and looking at the Rspec doc's, I'm not seeing anything else I need to require, include, or configure to get it to work. All the Rspec configure settings in Rails don't seem relevant (if I understand it correctly) as it has to do with database, fixtures, expect chaining, and order.

Rspec test:
```
require 'rspec/rails'
require 'rspec/autorun'

RSpec.describe "home page", :type => :request do
  it "subscribe_returns_success" do
    post '/subscribe', params: { device_token: "funny_token" }, headers: { "CONTENT_TYPE" => "application/json", "TOKEN" => "header_token" }
    expect(response.status).to eq(200)
  end

  it "unsubscribe_returns_success" do
    delete '/unsubscribe', params: { device_token: "funny_token" }, headers: { "CONTENT_TYPE" => "application/json", "TOKEN" => "header_token" }
    expect(response.status).to eq(200)
  end
end
```

Rspec Error:
```
I, [2016-05-22T10:43:53.919641 #2131]  INFO -- : Started POST "/subscribe" for 127.0.0.1 at 2016-05-22 10:43:53 -0700
D, [2016-05-22T10:43:53.920923 #2131] DEBUG -- : Error occurred while parsing request parameters.
Contents:

device_token=funny_token
F, [2016-05-22T10:43:53.923424 #2131] FATAL -- :   
F, [2016-05-22T10:43:53.923480 #2131] FATAL -- : ActionDispatch::ParamsParser::ParseError (822: unexpected token at 'device_token=funny_token'):
F, [2016-05-22T10:43:53.923494 #2131] FATAL -- :   
F, [2016-05-22T10:43:53.923505 #2131] FATAL -- : testcase.rb:76:in `block (2 levels) in <main>'

```

**Comment by sgrif** [comment]
If you cannot reproduce the error without using RSpec, it's a bug on their
end.

**Comment by hmistry** [comment]
@sgrif I would agree with your conclusion if I was at a good point where I have ruled out all the variables with confidence, but I'm not there yet.

Would you (or some one) help me get Rspec configured correctly with your above template script so I can do a `post` request without getting an `ActionDispatch` parsing error?

**Comment by rafaelfranca** [comment]
We are not Rspec users. You would find more help in the rspec-rails issue tracker. I'm happy to keep this open if we can isolate this problem to Rails but the fact that it is only reproducible with Rspec means that maybe it is not a Rails issue.

If you do a quick search in the Rails codebase by "needs to be", you will not see any error with that message so I'm sure that error is not being raised by Rails.

I'm more than happy to reopen if we are sure that this error is from Rails but right now we have more evidence that it is related to Rspec (or your application) than Rails.

**Comment by ajnye** [comment]
This definitely is an issue.  Not a bug in the code, but rather the documentation is wrong.

Based on [this page](http://api.rubyonrails.org/classes/ActionDispatch/Integration/RequestHelpers.html), `delete "/api/notifications/unsubscribe", params: { device_token: device.device_token }, headers: headers(token: token)` seems to be the correct way to write this request in a test.  However, [this](https://github.com/rails/rails/blob/9aaf8ca7672d525e1b5d54abc07ae9d589cf4e44/actionpack/test/controller/integration_test.rb) part of Rails' own source code tests shows that when using `params:` _and_ `headers:`, one must call `.to_json` as follows or one will get the error.

`      post "/foos_json.json", params: { foo: "fighters" }.to_json,
        headers: { "Content-Type" => "application/json" }`

**Comment by hmistry** [comment]
@ajnye It's good to know some one else had the same issue. I recall there was some inconsistent behaviors where sometimes I didn't need to use `to_json` and other times I did but didn't make the connection to what you found. Good to know and hope it gets fixed for consistency.

**Comment by ajnye** [comment]
Yeah. I have some more information about what I uncovered. I can write more about it and or try to fix the problem myself later.

Alexander Nye




On Thu, Sep 29, 2016 at 2:22 PM -0400, "hmistry" <notifications@github.com<mailto:notifications@github.com>> wrote:


@ajnye<https://github.com/ajnye> It's good to know some one else had the same issue. I recall there was some inconsistent behaviors where sometimes I didn't need to use to_json and other times I did but didn't make the connection to what you found. Good to know and hope it gets fixed for consistency.

-
